pkg/providers/v2/tags.go

```
@@ -0,0 +1,174 @@
+	// The tag key = TagNameKubernetesClusterPrefix + clusterID
+	// The tag value is an ownership value
+	TagNameKubernetesClusterPrefix = "kubernetes.io/cluster/"
+
+	// createTag* is configuration of exponential backoff for CreateTag call. We
+	// retry mainly because if we create an object, we cannot tag it until it is
+	// "fully created" (eventual consistency). Starting with 1 second, doubling
+	// it every step and taking 9 steps results in 255 second total waiting
+	// time.
+	createTagInitialDelay = 1 * time.Second
+	createTagFactor       = 2.0
+	createTagSteps        = 9
+)
+
+type awsTagging struct {
+	// ClusterName is our cluster identifier: we tag AWS resources with this value,
+	// and thus we can run two independent clusters in the same VPC or subnets.
+	ClusterName string
+}
+
+// newAWSTags is a constructor function for awsTagging
+func newAWSTags(clusterName string) (awsTagging, error) {
+	if clusterName != "" {
+		klog.Infof("AWS cloud filtering on ClusterName: %v", clusterName)
+	} else {
+		return awsTagging{}, errors.New("AWS cloud failed to find ClusterName")
+	}
+
+	return awsTagging{
+		ClusterName: clusterName,
+	}, nil
+}
+
+// Extracts the cluster name from the given tags, if they are present
+// If duplicate tags are found, returns an error
+func findClusterName(tags []*ec2.Tag) (string, error) {
+	clusterName := ""
+
+	for _, tag := range tags {
+		tagKey := aws.StringValue(tag.Key)
+		if strings.HasPrefix(tagKey, TagNameKubernetesClusterPrefix) {
+			name := strings.TrimPrefix(tagKey, TagNameKubernetesClusterPrefix)
+			if clusterName != "" {
+				return "", fmt.Errorf("Found multiple cluster tags with prefix %s (%q and %q)", TagNameKubernetesClusterPrefix, clusterName, name)
+			}
+			clusterName = name
+		}
+	}
+
+	return clusterName, nil
+}
+
+func (t *awsTagging) hasClusterTag(tags []*ec2.Tag) bool {
+	// if the clusterName is not configured -- we consider all instances.
```

Nit: wondering if this logic can be refactored slightly.  IMO returning true from a function called `hasClusterTag` when there isn't a tag to look for is bit confusing (for the sole purpose of causing some side effect behavior in the caller), wondering if its worth separating return values to have distinct values for "is cluster tag configured" and "does cluster tag exist".  We could return a specific error (or bool value) that the caller can check for if cluster tag is not configured.

---

good point, personally I prefer to return a specific error with some info for the caller to check, will update!